File: osint_nexus/core/detectors/tcp.py
from collections.abc import Mapping

from beartype import beartype

from osint_nexus.core.detectors.base import FingerprintStrategy
from osint_nexus.core.type_defs import JSONValue


class TcpFingerprintStrategy(FingerprintStrategy[Mapping[str, JSONValue], Mapping[str, JSONValue]]):
    """Strategy for TCP/IP stack fingerprinting (TTL/Window/Options)."""

    name: str = "tcp_stack"

    @beartype
    def extract(self, data: Mapping[str, JSONValue]) -> Mapping[str, JSONValue]:
        # Heuristic implementation, CC should be low (1 per block)
        ttl_value = data.get("ttl", 0)
        ttl = int(ttl_value) if isinstance(ttl_value, (int, float)) else 0

        # Safely extract tcp_options, ensuring it's a list of strings
        options_value = data.get("tcp_options")
        options: list[str] = []
        if isinstance(options_value, Mapping):  # Assuming it might be a Mapping if structure changed
            pass
        elif hasattr(options_value, "__iter__") and not isinstance(options_value, (str, bytes)):
            options = [str(v) for v in options_value if isinstance(v, (str, int, float, bool))]

        inferred_os, confidence = self._detect_os(ttl, options)

        return {
            "name": self.name,
            "data": {"inferred_os": inferred_os if inferred_os else "unknown"},
            "confidence": confidence,
        }

    def _detect_os(self, ttl: int, options: list[str]) -> tuple[str | None, float]:
        """Sub-method to maintain low CC."""
        if ttl == 128:
            return self._detect_windows(options)
        elif ttl == 64:
            return self._detect_linux_macos(options)
        elif ttl == 255:
            return "Network device (Cisco/Juniper)", 0.9
        return None, 0.0

    def _detect_windows(self, options: list[str]) -> tuple[str, float]:
        if "wscale" in options:
            return "Windows 10/11", 0.85
        return "Windows (older)", 0.7

    def _detect_linux_macos(self, options: list[str]) -> tuple[str, float]:
        if "timestamps" in options and "sack" in options:
            return "Linux (modern)", 0.75
        elif "timestamps" in options:
            return "macOS/iOS", 0.7
        return "Linux (older)", 0.5

File: internal/engine/strategies/tcp.go
package strategies

import (
	"context"
	"fmt"

	"github.com/FJ-cyberzilla/osint-nexus/internal/types"
)

// TCPPayload implements types.FingerprintPayload for TCP fingerprinting.
type TCPPayload struct {
	TTL        int      `json:"ttl"`
	TCPOptions []string `json:"tcp_options"`
}

// PayloadType returns the payload type identifier.
func (p TCPPayload) PayloadType() string { return "tcp_stack" }

// TCPOutputPayload implements types.FingerprintPayload for TCP fingerprinting output.
type TCPOutputPayload struct {
	InferredOS string `json:"inferred_os"`
}

// PayloadType returns the payload type identifier.
func (p TCPOutputPayload) PayloadType() string { return "tcp_stack_output" }

// TCPStrategy implements the FingerprintStrategy for TCP/IP stack fingerprinting (TTL/Window/Options).
type TCPStrategy struct{}

// NewTCPStrategy initializes a new TCPStrategy.
func NewTCPStrategy() *TCPStrategy {
	return &TCPStrategy{}
}

// Name returns the strategy name.
func (s *TCPStrategy) Name() string {
	return "tcp_stack"
}

// Extract performs TCP fingerprinting based on TTL and Options.
func (s *TCPStrategy) Extract(ctx context.Context, data types.FingerprintData) (types.FingerprintResult, error) {
	payload, ok := data.Payload.(TCPPayload)
	if !ok {
		return types.FingerprintResult{}, fmt.Errorf("tcp_strategy: invalid payload type: %T", data.Payload)
	}

	inferredOS, confidence := s.detectOS(payload.TTL, payload.TCPOptions)

	return types.FingerprintResult{
		Name: s.Name(),
		Data: types.FingerprintData{
			Type:    "tcp_stack",
			Payload: TCPOutputPayload{InferredOS: inferredOS},
		},
		Confidence: confidence,
	}, nil
}

func (s *TCPStrategy) detectOS(ttl int, options []string) (string, float64) {
	switch ttl {
	case 128:
		return s.detectWindows(options)
	case 64:
		return s.detectLinuxMacOS(options)
	case 255:
		return "Network device (Cisco/Juniper)", 0.9
	default:
		return "unknown", 0.0
	}
}

func (s *TCPStrategy) detectWindows(options []string) (string, float64) {
	for _, opt := range options {
		if opt == "wscale" {
			return "Windows 10/11", 0.85
		}
	}
	return "Windows (older)", 0.7
}

func (s *TCPStrategy) detectLinuxMacOS(options []string) (string, float64) {
	hasTimestamps := false
	hasSack := false
	for _, opt := range options {
		if opt == "timestamps" {
			hasTimestamps = true
		}
		if opt == "sack" {
			hasSack = true
		}
	}

	if hasTimestamps && hasSack {
		return "Linux (modern)", 0.75
	} else if hasTimestamps {
		return "macOS/iOS", 0.7
	}
	return "Linux (older)", 0.5
}
